language/move-vm/types/src/values/values_impl.rs

```diff
@@ -2066,113 +2042,80 @@ impl<'d> serde::de::DeserializeSeed<'d> for &Type {
-                }
-
-                macro_rules! deserialize_specialized_vec {
-                    ($tc: ident, $tc2: ident, $ty: ident) => {{
-                        struct $tc;
-                        impl<'d> serde::de::Visitor<'d> for $tc {
-                            type Value = Vec<$ty>;
-
-                            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
-                                formatter.write_str(stringify!($ty))
-                            }
-
-                            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
-                            where
-                                A: serde::de::SeqAccess<'d>,
-                            {
-                                let mut vals = Vec::new();
-                                while let Some(elem) = seq.next_element::<$ty>()? {
-                                    vals.push(elem)
-                                }
-                                Ok(vals)
-                            }
-                        }
-                        Ok(Value(ValueImpl::Container(Rc::new(RefCell::new(
-                            Container::$tc2(deserializer.deserialize_seq($tc)?),
-                        )))))
-                    }};
-                }
-
-                match &**layout {
-                    Type::U8 => deserialize_specialized_vec!(U8VectorVisitor, U8, u8),
-                    Type::U64 => deserialize_specialized_vec!(U64VectorVisitor, U64, u64),
-                    Type::U128 => deserialize_specialized_vec!(U128VectorVisitor, U128, u128),
-                    Type::Bool => deserialize_specialized_vec!(BoolVectorVisitor, Bool, bool),
-                    layout => Ok(Value(ValueImpl::Container(Rc::new(RefCell::new(
-                        deserializer.deserialize_seq(GeneralVectorVisitor(layout))?,
-                    ))))),
-                }
+                let container = match &**layout {
+                    Type::U8 => Container::U8(Vec::deserialize(deserializer)?),
+                    Type::U64 => Container::U64(Vec::deserialize(deserializer)?),
+                    Type::U128 => Container::U128(Vec::deserialize(deserializer)?),
+                    Type::Bool => Container::Bool(Vec::deserialize(deserializer)?),
+                    layout => Container::General(
+                        deserializer.deserialize_seq(VectorElementVisitor(layout))?,
+                    ),
+                };
+                Ok(Value(ValueImpl::Container(Rc::new(RefCell::new(
+                    container,
+                )))))
             }
 
-            Type::Struct(layout) => layout.deserialize(deserializer),
+            Type::Struct(ty) => {
+                let layout = &ty.layout;
+                let fields =
+                    deserializer.deserialize_tuple(layout.len(), StructFieldVisitor(layout))?;
+                Ok(Value::struct_(Struct::pack(fields)))
+            }
```

Comment:
I sort of think that this particular change is actually a little bit harder to read than using the `DeserializeSeed` impl

Reply:
Not sure I understand the comment. I just moved code around and didn't change the logic here.

Reply:
Specifically, `StructFieldVisitor(layout)` uses the types inside `layout` as seeds to deserialize the successive field values.

Reply:
How about putting it into an `impl DeserializeSeed for &StructType`? Currently only `Value` has `fn simple_deserialize` but that is not ideal. We only store resources on-chain so it makes more sense to use a StructType as the entry point.

Reply:
^ This is what I was tried to point out. Its nicer to imp DeserializeSeed than what the code currently does

Reply:
I see. @vgao1996 I was wondering if you wouldn't want to go into the opposite direction and get rid of Struct (de)serialization. It seems to be used in only one place, where it might be acceptable to deduce it from Value serialization (for the cost of one `layout.clone()`).

Reply:
Yeah that's exactly the problem: generally speaking, struct serialization/deserialization should be used instead of value serialization for type safety reasons. We are currently doing the opposite and we want to change that. Sorry for the confusion!

Reply:
ok let me update the code